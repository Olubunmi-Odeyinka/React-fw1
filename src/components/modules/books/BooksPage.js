import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import * as bookActions from '../../../actions/bookActions';
import ListView from "../../common/table/ListView";

class BooksPage extends React.Component {

   fieldsDefinition = {
      name: {
      },
      author: {
      },
      page: {
        type: 'number'
      },
      category: {
        type: 'number'
      }
    };

  componentDidMount =()=>{
    this.props.actions.loadBooks();
  }

  bookRow=(book, index)=> {
    return <div key={index}>{book.title}</div>;
  }

  redirectToAddBookPage=()=> {
    this.props.history.push('/book/create');
  }

  render=()=> {
    const {books} = this.props;

    return (
      <div className="container">
        <h1>Books</h1>
        <input type="submit"
               value="Add Book"
               className="btn btn-primary"
               onClick={this.redirectToAddBookPage}/>
        <ListView listItems={books} fieldsDefinition={this.fieldsDefinition}/>
      </div>
    );
  }
}

BooksPage.propTypes = {
  books: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired
};

function mapStateToProps(state, ownProps) {
  return {
    books: state.books,
    genres: state.genres
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(bookActions, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(BooksPage);
