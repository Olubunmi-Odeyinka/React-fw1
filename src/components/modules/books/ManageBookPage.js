import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {withFormik} from 'formik';
import Yup from 'yup';
import {bindActionCreators} from 'redux';
import * as bookActions from '../../../actions/bookActions';
import * as genreActions from '../../../actions/genreActions';
import GeneralForm from '../../common/form/GeneralForm';
import * as operation from '../../../constants/operationTypes';
import {genresFormattedForDropdown} from '../../common/selectors/selectors';
// import _ from 'underscore';
import toastr from 'toastr';
import uuid from 'uuid';

export class ManageBookPage extends React.Component {

  state = {
    formHeader: "Manage Books",
    formFieldReadOnly: false,
    urlOpeartion: '',
    listUrl:'/books',
    showSavingButton: true,
    savingString:['saving', 'save', 'btn-sucesss']
    // lookUps: {}
  };

    fieldsDefinition = {
        name: {
            type: 'text',
            label: 'Name'
        },
        author: {
            type: 'text',
            label: 'Author',
            focus: true
        },
        category: {
            type: 'select',
            label: 'Category'
        },
        page: {
            type: 'number',
            label: 'Book\'s Pages'
        }
    };

    componentDidMount =() =>{
      if (this.props.match.params.id) {
        this.props.actions.getBookById(this.props.match.params.id);
      }else{
        this.props.actions.loadBookSuccess({
          id: uuid(),
          name: 'Dada Ole',
          author: '',
          category: 4,
          page: ''
        });
      }

      if (this.props.match.params.operator) {

        this.setState(() => ({urlOpeartion : this.props.match.params.operator}));

        switch(this.props.match.params.operator){
          case operation.View_URL_String:
            this.setState(() => ({formFieldReadOnly : true}));
            this.setState(() => ({formHeader : "View Book"}));
            this.setState(() => ({showSavingButton : false}));
            break;
          case operation.Create_URL_String:
            this.setState(() => ({formHeader : "Create Book"}));
            break;
          case operation.Modify_URL_String:
            this.setState(() => ({formHeader : "Modify Book"}));
            this.setState(() => ({savingString : ['Updating', 'Update', 'btn-info']}));
            break;
          case operation.Delete_URL_String:
            this.setState(() => ({formHeader : "Delete Book"}));
            this.setState(() => ({formFieldReadOnly : true}));
            this.setState(() => ({savingString : ['Deleting', 'Delete', 'btn-danger']}));
            break;
          default:
            break;
        }
      }

      this.props.lookUpsAction.loadGenres();
    }

    redirect =(successMessage)=> {
        toastr.success(successMessage);
        this.props.history.push('/books');
        //Or this.context.router.history.push('/books');
    }

    FormDom = ({
                values,
                errors,
                touched,
                isSubmitting
            })=>(
        <GeneralForm
            formState={{...this.state}}
            lookUps={{'category':this.props.genres}}
            touched={touched}
            fields={this.fieldsDefinition}
            values={values}
            errors={errors}
            saving={isSubmitting}
        />
    );
   theReadonlyForm = withFormik({
     mapPropsToValues({book}){
       return{
         id: book.id,
         name: book.name,
         author: book.author,
         category: book.category,
         page: book.page
       }
     },
     enableReinitialize: true
   })(this.FormDom);

   theEditableForm = withFormik({
        mapPropsToValues({book}){
            return{
                id: book.id,
                name: book.name,
                author: book.author,
                category: book.category,
                page: book.page
            }
        },
        validationSchema: Yup.object().shape({
            name: Yup.string('Name must be a string').min(3, 'Name must be at least 3 characters').required('Name field is required'),
            author: Yup.string().max(10, 'Author can not be longer than 10').required('Author must be supplied')
        }),
        handleSubmit(values, bag){
            bag.setSubmitting(true);
            let subMissionPromise = null;
            let successMessage = null;
            switch(bag.props.urlOpeartion){
              case operation.Create_URL_String:
                subMissionPromise = bag.props.formAction.saveNewBook(values);
                successMessage = 'Book created';
                break;
              case operation.Modify_URL_String:
                subMissionPromise = bag.props.formAction.saveUpdateBook(values);
                successMessage = 'Book modified';
                break;
              case operation.Delete_URL_String:
                subMissionPromise = bag.props.formAction.saveDeleteBook(values);
                successMessage = 'Book deleted'
                break;
              default:
                subMissionPromise = bag.props.formAction.saveBook(values);
                successMessage = 'Book updated'
                break;
            }

           subMissionPromise
                .then(() => bag.props.redirect(successMessage))
                .catch(error => {
                    bag.setSubmitting(false);
                    toastr.error(error);
                });
        },
        enableReinitialize: true
    })(this.FormDom)

    render =()=> {
        return this.state.formFieldReadOnly ? (
          <this.theReadonlyForm
            book={this.props.book}
            redirect={this.redirect}
          />
        ): (
          <this.theEditableForm
            book={this.props.book}
            formAction={this.props.actions}
            urlOpeartion={this.state.urlOpeartion}
            redirect={this.redirect}
          />
        );
    }
}

ManageBookPage.propTypes = {
    book: PropTypes.object.isRequired,
    genres: PropTypes.array.isRequired,
    actions: PropTypes.object.isRequired
};

//Pull in the React Router context so router is available on this.context.router.
ManageBookPage.contextTypes = {
    router: PropTypes.object
};

function mapStateToProps(state, ownProps) {

    return {
        book: state.book,
        genres: genresFormattedForDropdown(state.genres)
    };
}

function mapDispatchToProps(dispatch) {
    return {
        lookUpsAction: bindActionCreators(genreActions, dispatch),
        actions: bindActionCreators(bookActions, dispatch),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(ManageBookPage);

