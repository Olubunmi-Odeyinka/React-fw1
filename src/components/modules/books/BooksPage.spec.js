import React from 'react';
import { shallow } from 'enzyme';
import { BooksPage } from './BooksPage';

import books from '../../../fixtures/books';
import genres from '../../../fixtures/genres';

let wrapper, props, itemUrl;

beforeEach(() => {
  props = {
    books,
    genres,
    actions:{
      loadBooks: jest.fn()
    }
  };
  itemUrl='';
});

test('should render BooksPage with non empty books', () => {
  wrapper = shallow(<BooksPage {...props} itemUrl={itemUrl}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render BooksPage with empty books', () => {
  props.books = [];
  wrapper = shallow(<BooksPage {...props} itemUrl={itemUrl}/>);
  expect(wrapper).toMatchSnapshot();
});

test('BooksPage should have a header Books', ()=>{
  wrapper = shallow(<BooksPage {...props} itemUrl={itemUrl}/>);
  expect(wrapper.find('h1').text()).toEqual('Books');
});

it('Create button is labeled Add Book', () => {
  wrapper = shallow(<BooksPage {...props} itemUrl={itemUrl}/>);
  expect(wrapper.find('input').props().value).toBe('Add Book');
});
