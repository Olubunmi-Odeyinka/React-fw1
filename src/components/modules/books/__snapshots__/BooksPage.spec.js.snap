// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BooksPage with empty books 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BooksPage
    actions={
        Object {
            "loadBooks": [Function],
          }
    }
    books={Array []}
    genres={
        Array [
            Object {
              "description": "That is an example",
              "id": 1,
              "name": "Rare Screen",
            },
            Object {
              "description": "5 The rate",
              "id": 2,
              "name": "Satire",
            },
            Object {
              "description": "This is new",
              "id": 3,
              "name": "Chain Taste",
            },
          ]
    }
    itemUrl=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Books
</h1>,
        <input
          className="btn btn-primary"
          onClick={[Function]}
          type="submit"
          value="Add Book"
/>,
        <ItemList
          fieldsDefinition={
                    Object {
                              "author": Object {},
                              "category": Object {
                                "type": "number",
                              },
                              "name": Object {},
                              "page": Object {
                                "type": "number",
                              },
                            }
          }
          itemUrl="/book"
          listItems={Array []}
/>,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Books",
        },
        "ref": null,
        "rendered": "Books",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "btn btn-primary",
          "onClick": [Function],
          "type": "submit",
          "value": "Add Book",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fieldsDefinition": Object {
            "author": Object {},
            "category": Object {
              "type": "number",
            },
            "name": Object {},
            "page": Object {
              "type": "number",
            },
          },
          "itemUrl": "/book",
          "listItems": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Books
</h1>,
          <input
            className="btn btn-primary"
            onClick={[Function]}
            type="submit"
            value="Add Book"
/>,
          <ItemList
            fieldsDefinition={
                        Object {
                                    "author": Object {},
                                    "category": Object {
                                      "type": "number",
                                    },
                                    "name": Object {},
                                    "page": Object {
                                      "type": "number",
                                    },
                                  }
            }
            itemUrl="/book"
            listItems={Array []}
/>,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Books",
          },
          "ref": null,
          "rendered": "Books",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "btn btn-primary",
            "onClick": [Function],
            "type": "submit",
            "value": "Add Book",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "fieldsDefinition": Object {
              "author": Object {},
              "category": Object {
                "type": "number",
              },
              "name": Object {},
              "page": Object {
                "type": "number",
              },
            },
            "itemUrl": "/book",
            "listItems": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`should render BooksPage with non empty books 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BooksPage
    actions={
        Object {
            "loadBooks": [Function],
          }
    }
    books={
        Array [
            Object {
              "author": "Russia Ban",
              "category": 3,
              "id": 1,
              "name": "The Syberia Story",
              "page": 456,
            },
            Object {
              "author": "Pact Sun",
              "category": 1,
              "id": 2,
              "name": "The Wolf Report",
              "page": 375,
            },
            Object {
              "author": "Brian Tracy",
              "category": 5,
              "id": 3,
              "name": "The Mayor of 3rd",
              "page": 590,
            },
          ]
    }
    genres={
        Array [
            Object {
              "description": "That is an example",
              "id": 1,
              "name": "Rare Screen",
            },
            Object {
              "description": "5 The rate",
              "id": 2,
              "name": "Satire",
            },
            Object {
              "description": "This is new",
              "id": 3,
              "name": "Chain Taste",
            },
          ]
    }
    itemUrl=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Books
</h1>,
        <input
          className="btn btn-primary"
          onClick={[Function]}
          type="submit"
          value="Add Book"
/>,
        <ItemList
          fieldsDefinition={
                    Object {
                              "author": Object {},
                              "category": Object {
                                "type": "number",
                              },
                              "name": Object {},
                              "page": Object {
                                "type": "number",
                              },
                            }
          }
          itemUrl="/book"
          listItems={
                    Array [
                              Object {
                                "author": "Russia Ban",
                                "category": 3,
                                "id": 1,
                                "name": "The Syberia Story",
                                "page": 456,
                              },
                              Object {
                                "author": "Pact Sun",
                                "category": 1,
                                "id": 2,
                                "name": "The Wolf Report",
                                "page": 375,
                              },
                              Object {
                                "author": "Brian Tracy",
                                "category": 5,
                                "id": 3,
                                "name": "The Mayor of 3rd",
                                "page": 590,
                              },
                            ]
          }
/>,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Books",
        },
        "ref": null,
        "rendered": "Books",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "btn btn-primary",
          "onClick": [Function],
          "type": "submit",
          "value": "Add Book",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fieldsDefinition": Object {
            "author": Object {},
            "category": Object {
              "type": "number",
            },
            "name": Object {},
            "page": Object {
              "type": "number",
            },
          },
          "itemUrl": "/book",
          "listItems": Array [
            Object {
              "author": "Russia Ban",
              "category": 3,
              "id": 1,
              "name": "The Syberia Story",
              "page": 456,
            },
            Object {
              "author": "Pact Sun",
              "category": 1,
              "id": 2,
              "name": "The Wolf Report",
              "page": 375,
            },
            Object {
              "author": "Brian Tracy",
              "category": 5,
              "id": 3,
              "name": "The Mayor of 3rd",
              "page": 590,
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Books
</h1>,
          <input
            className="btn btn-primary"
            onClick={[Function]}
            type="submit"
            value="Add Book"
/>,
          <ItemList
            fieldsDefinition={
                        Object {
                                    "author": Object {},
                                    "category": Object {
                                      "type": "number",
                                    },
                                    "name": Object {},
                                    "page": Object {
                                      "type": "number",
                                    },
                                  }
            }
            itemUrl="/book"
            listItems={
                        Array [
                                    Object {
                                      "author": "Russia Ban",
                                      "category": 3,
                                      "id": 1,
                                      "name": "The Syberia Story",
                                      "page": 456,
                                    },
                                    Object {
                                      "author": "Pact Sun",
                                      "category": 1,
                                      "id": 2,
                                      "name": "The Wolf Report",
                                      "page": 375,
                                    },
                                    Object {
                                      "author": "Brian Tracy",
                                      "category": 5,
                                      "id": 3,
                                      "name": "The Mayor of 3rd",
                                      "page": 590,
                                    },
                                  ]
            }
/>,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Books",
          },
          "ref": null,
          "rendered": "Books",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "btn btn-primary",
            "onClick": [Function],
            "type": "submit",
            "value": "Add Book",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "fieldsDefinition": Object {
              "author": Object {},
              "category": Object {
                "type": "number",
              },
              "name": Object {},
              "page": Object {
                "type": "number",
              },
            },
            "itemUrl": "/book",
            "listItems": Array [
              Object {
                "author": "Russia Ban",
                "category": 3,
                "id": 1,
                "name": "The Syberia Story",
                "page": 456,
              },
              Object {
                "author": "Pact Sun",
                "category": 1,
                "id": 2,
                "name": "The Wolf Report",
                "page": 375,
              },
              Object {
                "author": "Brian Tracy",
                "category": 5,
                "id": 3,
                "name": "The Mayor of 3rd",
                "page": 590,
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
